compiler/qsc_linter/src/lints/hir.rs

```diff
@@ -29,7 +35,9 @@ use super::lint;
 //  For more details on how to add a lint, please refer to the crate-level documentation
 //  in qsc_linter/lib.rs.
 declare_hir_lints! {
-    (NeedlessOperation, LintLevel::Allow, "operation does not contain any quantum operations", "this callable can be declared as a function instead")
+    (NeedlessOperation, LintLevel::Allow, "operation does not contain any quantum operations", "this callable can be declared as a function instead"),
+    (DeprecatedFunctionConstructor, LintLevel::Warn, "deprecated function constructors", "function constructors for struct types are deprecated, use `new` instead"),
+    (DeprecatedWithOperator, LintLevel::Warn, "deprecated `w/` and `w/=` operators for structs", "`w/` and `w/=` operators for structs are deprecated, use `new` instead", Option<WithOperatorLint>),
```

Comment:
As we discussed yesterday, we shouldn't put implementation details of the lints such as `WithOperatorLint` here. This macro is intended to hold metadata about the lints such as their names, default levels and messages.

Fitting this implementation detail into the macro also introduces an extra layer of indirection to access this auxiliary information that is not evident that exists, `self.lint_info`.

Reply:
I wanted to use the macro to remove boilerplate code for each of the lints. This change allows the structs to be declared by the macro and includes the necessary `level` field. Only lints that need lint-specific information will have the additional `info` field whose type corresponds to the given last parameter of the macro.

These changes make it easier to write new lints, both those that don't need extra data and those that do. If it is a matter of communicating that in code, the description in the comments can be updated to make that clearer. I also think that this is a smoother experience than requiring the declaration of structs that have a required `level` field, as when they don't have that there are obfuscated error messages due to the macro.

If you feel strongly about this change, then I will defer to you as the code-owner, but I _do_ think that these changes make for a better coding experience and ease of maintenance.

Reply:
I've reverted the commit in question, removing the changes to the macro.